pkg/customresourcestate/generate/generator/generator.go

```diff
@@ -0,0 +1,140 @@
+/*
+Copyright 2023 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package generator
+
+import (
+	"fmt"
+	"sort"
+
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-tools/pkg/crd"
+	"sigs.k8s.io/controller-tools/pkg/genall"
+	"sigs.k8s.io/controller-tools/pkg/loader"
+	ctrlmarkers "sigs.k8s.io/controller-tools/pkg/markers"
+
+	"k8s.io/kube-state-metrics/v2/pkg/customresourcestate"
+	"k8s.io/kube-state-metrics/v2/pkg/customresourcestate/generate/markers"
+)
+
+// CustomResourceConfigGenerator implements the Generator interface from controller-tools.
+// It uses markers to generate a custom resource configuration for kube-state-metrics from go code.
+type CustomResourceConfigGenerator struct{}
+
+var _ genall.Generator = &CustomResourceConfigGenerator{}
+var _ genall.NeedsTypeChecking = &CustomResourceConfigGenerator{}
+
+// RegisterMarkers registers all markers needed by this Generator
+// into the given registry.
+func (g CustomResourceConfigGenerator) RegisterMarkers(into *ctrlmarkers.Registry) error {
+	for _, m := range markers.MarkerDefinitions {
+		if err := m.Register(into); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Generate generates artifacts produced by this marker.
+// It's called after RegisterMarkers has been called.
+func (g CustomResourceConfigGenerator) Generate(ctx *genall.GenerationContext) error {
+	// Create the parser which is specific to the metric generator.
+	parser := newParser(
+		&crd.Parser{
+			Collector: ctx.Collector,
+			Checker:   ctx.Checker,
+		},
+	)
+
+	// Loop over all passed packages.
+	for _, pkg := range ctx.Roots {
+		// skip packages which don't import metav1 because they can't define a CRD without meta v1.
+		metav1 := pkg.Imports()["k8s.io/apimachinery/pkg/apis/meta/v1"]
+		if metav1 == nil {
```

Comment:
Should we have debug logging here to inform which ones were skipped? Could make it easier for users to figure out.

Reply:
Happy to add a log line, but I think this could be misleading.
There are lots of other cases where a package gets skipped, but thats inside controller-tools's `crd.FindKubeKinds` which we call [here](https://github.com/kubernetes/kube-state-metrics/blob/efe2cac20b9d118a7c0daeb72368cbaff2fa9560/pkg/customresourcestate/generate/generator/generator.go#L73):

https://github.com/kubernetes-sigs/controller-tools/blob/v0.13.0/pkg/crd/gen.go#L238